20230729-db-cloud-auth.md

```diff
@@ -0,0 +1,172 @@
+# Meta
+[meta]: #meta
+- Name: Cloud Authentication for database access
+- Start Date: 2030-07-29
+- Author(s): [KlausVii](https://github.com/KlausVii)
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- Relevant Issues: https://github.com/openfga/openfga/issues/870
+- Supersedes: "N/A"
+
+# Summary
+[summary]: #summary
+
+Many users run OpenFGA in a cloud environment, and these cloud providers often provide alternate authentication methods for their databases,
+such as [AWS RDS IAM](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html) and [GCP Cloud SQL Go Connector](https://cloud.google.com/sql/docs/postgres/samples/cloud-sql-postgres-databasesql-connect-connector).
+This RFC proposes a new authentication method that allows users to authenticate against a cloud provider, starting with AWS RDS support.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+  - Enabling cloud authentication methods will allow users to deploy OpenFGA faster and more securely.
+  - Personally at our workplace all our other services authenticate with RDS IAM and not being able to do so with OpenFGA adds friction.
+  - I imagine there are similar issues [on other cloud providers](https://cloud.google.com/sql/docs/mysql/connect-admin-ip#connect-ssl).
+- What use cases does it support?
+  - Users will be able to use their cloud provider's authentication method to authenticate against their database. 
+- What is the expected outcome?
+  - openFGA supports AWS IAM.
+  - it is easy to extend to other cloud providers. 
+
+# What it is
+[what-it-is]: #what-it-is
+
+Add support for cloud provider authentication methods to openFGA, so that operators can more easily deploy openFGA in a cloud environment.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## New configuration options
+
+The most straight forward option would be to introduce a new `datastore` string option `auth_method`. The user could then 
+select from a list of supported authentication methods. If left blank, the default would be to use the existing username and password 
+specified by the URI.
+
+`-datastore.auth_method (string): "aws_rds_iam" | "gcp_cloud_iam" | etc`
+
+Alternatively, a set of boolean flags could be used, but that seems cumbersome since only one would ever be used at a time.
+
+`-datastore.aws_rds_iam_auth (bool)`
+
+#### Inclusion of cloud provider SDKs configuration
+
+Including the provider specific options, like `AWS_REGION`, seems unnecessary since the SDK's themselves are able to 
+aqcuire these from the environment automatically. And adding them would make the configuration more complex and dependent on the SDKs. 
+
+## Implementation
+
+Authentication would need to be implemented for both databases supported by openFGA: postgres and mysql.
```

Comment:
As a maintainer team we're worried about introducing these cloud specific database connection functionalities in OpenFGA, because it will be cumbersome to manage. Depending on more of a community driven and community based standard for this functionality would be preferable as it allows our maintainer team to leverage a lager community to help evolve and maintain these cloud IAM mechanisms.

For this reason, we'd like to see an exploration of the `go-cloud` CDK (mentioned in the Prior Art section below) before we proceed forward. Specifically,

* Are there features/functionalities that the Go Cloud CDK does not address that are pertinent to the usage patterns for these proprietary cloud IAM protocols? If so, what are the limitations? For example, can you use the assumed role functionality of AWS IAM etc..

* Does the usage of the cloud specific sql providers in the Go Cloud CDK necessitate a specific underlying driver (e.g. pgx, pql, etc..)? Or can you bring your own driver?

  From the looks of it you can use whatever underlying driver you want (see https://github.com/google/go-cloud/blob/master/samples/guestbook/wire_gen.go#L16C34-L16C34 as an example)

If you'd be willing to help with this discovery @KlausVii we'd really appreciate it. If not though, our next step as a maintainer team is to do some more discovery with the go-cloud SDK and see how it works for this use case.

We have used the go-cloud SDK elsewhere and have had positive experiences with it, so we have at least some confidence in it as a potential solution.